File: LCOF/21-30/22/22.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def getKthFromEnd(self, head, k):
        """
        :type head: ListNode
        :type k: int
        :rtype: ListNode
        """

        stack = []
        node = head
        while node:
            stack.append(node)
            node = node.next

        result = None
        for _ in range(k):
            result = stack.pop()

        return result

    def getKthFromEnd2(self, head, k):
        """
        :type head: ListNode
        :type k: int
        :rtype: ListNode
        """

        fast = slow = head
        for _ in range(k):
            fast = fast.next

        while fast:
            fast = fast.next
            slow = slow.next

        return slow

File: LCOF/21-30/22/22.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */

struct ListNode* getKthFromEnd(struct ListNode* head, int k){
    struct ListNode *fast = head, *slow = head;
    int i = 0;
    while (i++ < k)
        fast = fast->next;

    while (fast) {
        fast = fast->next;
        slow = slow->next;
    }

    return slow;
}